SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py

```diff
@@ -0,0 +1,86 @@
+from typing import Optional
+from math import exp
+
+from SC_redesign.Crop_and_Soil.soil.soil_data import SoilData
+
+"""
+This module handles the denitrification of nitrogen in the nitrates pool, based on SWAT section 3:1.4.
+"""
+
+
+class Denitrification:
+
+    def __init__(self, soil_data: Optional[SoilData] = None, field_size: Optional[float] = None):
+        """This method initializes the SoilData object that this module will work with, or create one if none provided.
+
+        Parameters
+        ----------
+        soil_data : SoilData, optional
+            The SoilData object used by this module to track the denitrification of nitrates in the soil profile,
+            creates new one if one is not provided.
+        field_size : float, optional
+            Size of the field (ha)
+
+        Notes
+        -----
+        The field size is used to initialize a SoilData object for this module to work with, if a pre-configured
+        SoilData object is not provided.
+
+        """
+        self.data = soil_data or SoilData(field_size=field_size)
+
+    def do_denitrification(self) -> None:
+        """Conducts the daily denitrification operations.
+
+        References
+        ----------
+        SWAT Theoretical documentation section 3:1.4, eqn. 3:1.4.1, 2
+
+        """
+        for layer in self.data.soil_layers:
+            if layer.nutrient_cycling_water_factor < layer.denitrification_threshold_water_content:
+                continue
+
+            # TODO: check that soil_overall_carbon_fraction is being used correctly in this context - issue #475
+            nitrified_nitrates = self._calculate_nitrification_amount(layer.nitrate_content,
+                                                                      layer.denitrification_rate_coefficient,
+                                                                      layer.nutrient_cycling_temp_factor,
+                                                                      layer.soil_overall_carbon_fraction * 100)
+            layer.nitrate_content -= nitrified_nitrates
+            layer.annual_denitrified_nitrogen_total += nitrified_nitrates
+
+    # --- Static methods ---
+    @staticmethod
+    def _calculate_nitrification_amount(nitrate_content: float, denitrification_rate_coefficient: float,
+                                        temp_factor: float, percent_organic_carbon_content: float) -> float:
+        """This method calculates the amount of nitrate lost to denitrification.
```

Comment:
would it make sense to bring this here?
```suggestion
        """Calculates the fraction of nitrates lost to nitrification as nitrification factor, and bounds it to be in the range [0.0, 1.0].
```
maybe explain how bounding works? do we clamp or normalize?

Reply:
I like keeping comments about specific implementation details in the `Notes` section, IMO it keeps the docstring a bit cleaner. Also this method calculates the amount of nitrate lost, calculating the nitrification factor is one of the components of this process.

I am not sure what the difference between normalizing and clamping is (but if I had to guess, I'd say clamp). As for an explanation of why the bounds are what they are, I think it would be pretty obvious to a SME but adding a note wouldn't hurt.